In [1]:
#settings
import pandas as pd
import sqlite3
pd.set_option('max_columns', 180)
pd.set_option('max_rows', 200000)
pd.set_option('max_colwidth', 5000)

In [2]:
games = pd.read_csv('game_log.csv')
print(games.shape)
games.head()

C:\Users\Adam\Miniconda3\envs\minimal_ds\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (12,13,14,15,19,20,81,82,83,84,85,86,87,88,93,94,95,96,97,98,99,100,105,106,108,109,111,112,114,115,117,118,120,121,123,124,126,127,129,130,132,133,135,136,138,139,141,142,144,145,147,148,150,151,153,154,156,157,160) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(171907, 161)


date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1      NaN              1    FW1   
1  18710505               0         Fri    BS1      NaN              1    WS3   
2  18710506               0         Sat    CL1      NaN              2    RC1   
3  18710508               0         Mon    CL1      NaN              3    CH1   
4  18710509               0         Tue    BS1      NaN              2    TRO   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0      NaN              1        0        2         54.0         D        NaN   
1      NaN              1       20       18         54.0         D        NaN   
2      NaN              1       12        4         54.0         D        NaN   
3      NaN              1       12       14         54.0         D        NaN   
4      NaN              1        9        5         54.0         D        NaN   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0      NaN     NaN   FOR01       200.0           120.0    000000000   
1      NaN     NaN   WAS01      5000.0           145.0    107000435   
2      NaN     NaN   RCK01      1000.0           140.0    610020003   
3      NaN     NaN   CHI01      5000.0           150.0    101403111   
4      NaN     NaN   TRO01      3250.0           145.0    000002232   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   
3    077000000       46.0    15.0        2.0        1.0         2.0   10.0   
4    101003000       46.0    17.0        4.0        1.0         0.0    6.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   
3               0.0                0.0             0.0      0.0   
4               0.0                0.0             0.0      2.0   

   v_intentional_walks  v_strikeouts  v_stolen_bases  v_caught_stealing  \
0                  NaN           6.0             1.0                NaN   
1                  NaN           5.0             3.0                NaN   
2                  NaN           1.0             0.0                NaN   
3                  NaN           1.0             0.0                NaN   
4                  NaN           0.0             1.0                NaN   

   v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
0                    -1.0                           NaN             4.0   
1                    -1.0                           NaN            12.0   
2                    -1.0                           NaN            10.0   
3                    -1.0                           NaN             7.0   
4                    -1.0                           NaN            12.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   
3              1.0                       6.0                 6.0   
4              1.0                       2.0                 2.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   
3             0.0      0.0       27.0       15.0      11.0             6.0   
4             0.0      0.0       

The game database contains a wealth of information (161 columns!) for over 170k baseball games (rows) going back to May, 1871. Data for individual games include home and away teams, scores, aggregate statistics for each team, pitchers, lineups, umpires, managers and more!

In [3]:
parks = pd.read_csv('park_codes.csv')
print(parks.shape)
parks.head()

(252, 9)


park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN

The parks database contains 9 columns of data for 252 big league parks. Included are names, locations, and start and end dates for it's use as a ballpark. It is linked to the game logs by a park_id column. 

In [4]:
people = pd.read_csv('person_codes.csv')
print(people.shape)
people.head()

(20494, 7)


id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004       NaN         NaN       NaN
1  aaroh101    Aaron    Hank   04/13/1954       NaN         NaN       NaN
2  aarot101    Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3  aased001     Aase     Don   07/26/1977       NaN         NaN       NaN
4  abada001     Abad    Andy   09/10/2001       NaN         NaN       NaN

The people database contains data about people's debut and role in MLB. Each person has a unique id - the id column - that links them to fields in the game logs.

In [5]:
teams = pd.read_csv('team_codes.csv')
print(teams.shape)
teams.head()

(150, 8)


team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1    NaN   1872  1874  Baltimore         Canaries       BL1    1

The teams database contains data for 149 teams that have participated in MLB, including start and end dates, city, and nicknames. The teams table is tied to the games log by the team_id which is represented by the home or visiting team columns in the game logs.

In [6]:
# Create database and helper functions

DB = "mlb.db"

def run_query(q):
    with sqlite3.connect(DB) as conn:
        return pd.read_sql(q,conn)

def run_command(c):
    with sqlite3.connect(DB) as conn:
        conn.execute('PRAGMA foreign_keys = ON;')
        conn.isolation_level = None
        conn.execute(c)

def show_tables():
    q = '''
    SELECT
        name,
        type
    FROM sqlite_master
    WHERE type IN ("table","view");
    '''
    return run_query(q)

tables = {
    "game_log": games,
    "person_codes": people,
    "team_codes": teams,
    "park_codes": parks
}

with sqlite3.connect(DB) as conn:    
    for name, data in tables.items():
        conn.execute("DROP TABLE IF EXISTS {};".format(name))
        data.to_sql(name,conn,index=False)
        
show_tables()

name   type
0              park  table
1            league  table
2            person  table
3              team  table
4              game  table
5   team_appearance  table
6   appearance_type  table
7          game_log  table
8      person_codes  table
9        team_codes  table
10       park_codes  table

In [7]:
# Add primary key

c1 = '''
ALTER TABLE game_log
ADD COLUMN game_id TEXT;
'''

try:
    run_command(c1)
except:
    pass

c2 = '''
UPDATE game_log
SET game_id = h_name || date || number_of_game
WHERE game_id is NULL;
'''

run_command(c2)

q1 = '''
SELECT 
    h_name,
    date,
    number_of_game,
    game_id
FROM game_log
LIMIT 5;
'''

run_query(q1)

h_name      date  number_of_game       game_id
0    FW1  18710504               0  FW1187105040
1    WS3  18710505               0  WS3187105050
2    RC1  18710506               0  RC1187105060
3    CH1  18710508               0  CH1187105080
4    TRO  18710509               0  TRO187105090

Now that we've created our primary key for main dataset it is time to look for normalization opportunities. First we'll identify duplicate and redundant data:

* All of the names (players, managers, umpires, etc.) in the game_log table are redundant since that information already exists in the people table. The two tables are linked by the person id.
* The start, end, and league fields in the park table are redundant. That information already exists in the game_log table.
* The debut fields in the person table are redundant as each person's debut in any role they filled is already in the game_log table.
* In the teams table, the league, start, and end dates are all available in the game_log table.
* There are a lot of duplicate columns in the game log table - statistics and player positions for each team for example. Creating a new table with seperate entries for the home and away team would dramatically cut down on these.
* Creating a seperate table for apperance type and linking it to game_id would drastically cut down on duplicate entries and simplify the main games table without the loss of any information.

Given the above information, the schema we'll use for the rest of this project is included below:

![mlb_db_design](mlb_db_design.png)

We'll now create the tables to match the above schema.

In [8]:
# Create perosn table

c3 = '''
CREATE TABLE IF NOT EXISTS person (
person_id TEXT PRIMARY KEY,
first_name TEXT,
last_name TEXT);
'''

c4 = '''
INSERT OR IGNORE INTO person
SELECT
    id,
    first,
    last
FROM person_codes;
'''

q2 = '''
SELECT * FROM person
LIMIT 5;
'''

run_command(c3)
run_command(c4)
run_query(q2)

person_id first_name last_name
0  aardd001      David   Aardsma
1  aaroh101       Hank     Aaron
2  aarot101     Tommie     Aaron
3  aased001        Don      Aase
4  abada001       Andy      Abad

In [9]:
# Create park table

c5 = '''
CREATE TABLE IF NOT EXISTS park (
park_id TEXT PRIMARY KEY,
name TEXT,
nickname TEXT,
city TEXT,
state TEXT,
notes TEXT);
'''

c6 = '''
INSERT OR IGNORE INTO park
SELECT
    park_id,
    name,
    aka,
    city,
    state,
    notes
FROM park_codes;
'''

q3 = '''
SELECT * FROM park
LIMIT 5;'''

run_command(c5)
run_command(c6)
run_query(q3)

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                   nickname       city state  \
0                                      None     Albany    NY   
1                                      None    Altoona    PA   
2             Edison Field; Anaheim Stadium    Anaheim    CA   
3                                      None  Arlington    TX   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX   

                                            notes  
0  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                            None  
2                                            None  
3                                            None  
4                                            None

In [10]:
# Create league table

c7 = '''
CREATE TABLE IF NOT EXISTS league (
league_id TEXT PRIMARY KEY,
name TEXT);
'''

c8 = '''
INSERT OR IGNORE INTO league
VALUES
    ('NL', 'National League'),
    ('AL', 'American League'),
    ('AA', 'American Association'),
    ('FL', 'Federal League'),
    ('PL', 'Players League'),
    ('UA', 'Union Association')
;
'''

q4 = '''
SELECT * FROM league
LIMIT 6;'''

run_command(c7)
run_command(c8)
run_query(q4)

league_id                  name
0        NL       National League
1        AL       American League
2        AA  American Association
3        FL        Federal League
4        PL        Players League
5        UA     Union Association

In [11]:
# Create apperance type table

c9 = '''
DROP TABLE IF EXISTS appearance_type;
'''

c10 = '''
CREATE TABLE appearance_type (
appearance_type_id TEXT PRIMARY KEY,
name TEXT,
category TEXT);
'''

run_command(c9)
run_command(c10)

appearance_type = pd.read_csv('appearance_type.csv')
with sqlite3.connect(DB) as conn:
    appearance_type.to_sql('appearance_type', conn, index=False, if_exists='append')

q5 = '''
SELECT * FROM appearance_type
'''

run_query(q5)

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [12]:
# Create team table

c11 = '''
CREATE TABLE IF NOT EXISTS team (
team_id TEXT PRIMARY KEY,
league_id TEXT,
city TEXT,
nickname TEXT,
franch_id TEXT,
FOREIGN KEY (league_id) REFERENCES league(league_id)
);
'''

c12 = '''
INSERT OR IGNORE INTO team
SELECT
    team_id,
    league,
    city,
    nickname,
    franch_id
FROM team_codes;
'''

q6 = '''
SELECT * FROM team
LIMIT 5;
'''

run_command(c11)
run_command(c12)
run_query(q6)

team_id league_id       city         nickname franch_id
0     ALT        UA    Altoona  Mountain Cities       ALT
1     ARI        NL    Arizona     Diamondbacks       ARI
2     BFN        NL    Buffalo           Bisons       BFN
3     BFP        PL    Buffalo           Bisons       BFP
4     BL1      None  Baltimore         Canaries       BL1

In [13]:
# Create game table

c13 = '''
CREATE TABLE IF NOT EXISTS game (
game_id TEXT PRIMARY KEY,
date TEXT,
number_of_game INTEGER,
park_id TEXT,
length_outs FLOAT,
day TEXT,
completion INTEGER,
forefeit INTEGER,
protest INTEGER,
attendance INTEGER,
length_minutes INTEGER,
additional_info TEXT,
acquisition_info TEXT,
FOREIGN KEY (park_id) REFERENCES park(park_id)
);'''

c14 = '''
INSERT OR IGNORE into game
SELECT
    game_id,
    date,
    number_of_game,
    park_id,
    length_outs,
    day_of_week,
    completion,
    forefeit,
    protest,
    attendance,
    length_minutes,
    additional_info,
    acquisition_info
FROM game_log;
'''

q7 = '''
SELECT * FROM game
LIMIT 5;
'''

run_command(c13)
run_command(c14)
run_query(q7)

game_id      date  number_of_game park_id  length_outs  day  \
0  FW1187105040  18710504               0   FOR01         54.0  Thu   
1  WS3187105050  18710505               0   WAS01         54.0  Fri   
2  RC1187105060  18710506               0   RCK01         54.0  Sat   
3  CH1187105080  18710508               0   CHI01         54.0  Mon   
4  TRO187105090  18710509               0   TRO01         54.0  Tue   

  completion fofeit protest  attendance  length_minutes additional_info  \
0       None   None    None         200             120            None   
1       None   None    None        5000             145            HTBF   
2       None   None    None        1000             140            None   
3       None   None    None        5000             150            None   
4       None   None    None        3250             145            HTBF   

  acquisition_info  
0                Y  
1                Y  
2                Y  
3                Y  
4                Y

In [14]:
# Create team appearance table

c15 = '''
CREATE TABLE IF NOT EXISTS team_appearance (
    team_id TEXT,
    game_id TEXT,
    home BOOLEAN,
    league_id TEXT,
    score INTEGER,
    line_score TEXT,
    at_bats INTEGER,
    hits INTEGER,
    doubles INTEGER,
    triples INTEGER,
    homeruns INTEGER,
    rbi INTEGER,
    sacrifice_hits INTEGER,
    sacrifice_flies INTEGER,
    hit_by_pitch INTEGER,
    walks INTEGER,
    intentional_walks INTEGER,
    strikeouts INTEGER,
    stolen_bases INTEGER,
    caught_stealing INTEGER,
    grounded_into_double INTEGER,
    first_catcher_interference INTEGER,
    left_on_base INTEGER,
    pitchers_used INTEGER,
    individual_earned_runs INTEGER,
    team_earned_runs INTEGER,
    wild_pitches INTEGER,
    balks INTEGER,
    putouts INTEGER,
    assists INTEGER,
    errors INTEGER,
    passed_balls INTEGER,
    double_plays INTEGER,
    triple_plays INTEGER,
    PRIMARY KEY (team_id, game_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id)
);'''

c16 = '''
INSERT OR IGNORE INTO team_appearance
    SELECT
        h_name,
        game_id,
        1 AS home,
        h_league,
        h_score,
        h_line_score,
        h_at_bats,
        h_hits,
        h_doubles,
        h_triples,
        h_homeruns,
        h_rbi,
        h_sacrifice_hits,
        h_sacrifice_flies,
        h_hit_by_pitch,
        h_walks,
        h_intentional_walks,
        h_strikeouts,
        h_stolen_bases,
        h_caught_stealing,
        h_grounded_into_double,
        h_first_catcher_interference,
        h_left_on_base,
        h_pitchers_used,
        h_individual_earned_runs,
        h_team_earned_runs,
        h_wild_pitches,
        h_balks,
        h_putouts,
        h_assists,
        h_errors,
        h_passed_balls,
        h_double_plays,
        h_triple_plays
    FROM game_log

UNION

    SELECT    
        v_name,
        game_id,
        0 AS home,
        v_league,
        v_score,
        v_line_score,
        v_at_bats,
        v_hits,
        v_doubles,
        v_triples,
        v_homeruns,
        v_rbi,
        v_sacrifice_hits,
        v_sacrifice_flies,
        v_hit_by_pitch,
        v_walks,
        v_intentional_walks,
        v_strikeouts,
        v_stolen_bases,
        v_caught_stealing,
        v_grounded_into_double,
        v_first_catcher_interference,
        v_left_on_base,
        v_pitchers_used,
        v_individual_earned_runs,
        v_team_earned_runs,
        v_wild_pitches,
        v_balks,
        v_putouts,
        v_assists,
        v_errors,
        v_passed_balls,
        v_double_plays,
        v_triple_plays
    from game_log;
'''

q8 = """
SELECT * FROM team_appearance
WHERE game_id = (
                 SELECT MIN(game_id) from game
                )
   OR game_id = (
                 SELECT MAX(game_id) from game
                )
ORDER By game_id, home;
"""
run_command(c15)
run_command(c16)
run_query(q8)

team_id       game_id  home league_id  score line_score at_bats  hits  \
0     SLU  ALT188404300     0        UA     15       None    None  None   
1     ALT  ALT188404300     1        UA      2       None    None  None   
2     BSU  WSU188409250     0        UA      2       None    None  None   
3     WSU  WSU188409250     1        UA     10       None    None  None   

  doubles triples homeruns   rbi sacrifice_hits sacrifice_flies hit_by_pitch  \
0    None    None     None  None           None            None         None   
1    None    None     None  None           None            None         None   
2    None    None     None  None           None            None         None   
3    None    None     None  None           None            None         None   

  walks intentional_walks strikeouts stolen_bases caught_stealing  \
0  None              None       None         None            None   
1  None              None       None         None            None   
2  None              None       None         None            None   
3  None              None       None         None            None   

  grounded_into_double first_catcher_interference left_on_base pitchers_used  \
0                 None                       None         None          None   
1                 None                       None         None          None   
2                 None                       None         None          None   
3                 None                       None         None          None   

  individual_earned_runs team_earned_runs wild_pitches balks putouts assists  \
0                   None             None         None  None    None    None   
1                   None             None         None  None    None    None   
2                   None             None         None  None    None    None   
3                   None             None         None  None    None    None   

  errors passed_balls double_plays triple_plays  
0   None         None         None         None  
1   None         None         None         None  
2   None         None         None         None  
3   None         None         None         None

In [15]:
# Create person appearance table

c17 = "DROP TABLE IF EXISTS person_appearance"

run_command(c17)

c18 = """
CREATE TABLE person_appearance (
    appearance_id INTEGER PRIMARY KEY,
    person_id TEXT,
    team_id TEXT,
    game_id TEXT,
    appearance_type_id,
    FOREIGN KEY (person_id) REFERENCES person(person_id),
    FOREIGN KEY (team_id) REFERENCES team(team_id),
    FOREIGN KEY (game_id) REFERENCES game(game_id),
    FOREIGN KEY (appearance_type_id) REFERENCES appearance_type(appearance_type_id)
);
"""

c19 = """
INSERT OR IGNORE INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        NULL,
        hp_umpire_id,
        "UHP"
    FROM game_log
    WHERE hp_umpire_id IS NOT NULL    

UNION

    SELECT
        game_id,
        NULL,
        [1b_umpire_id],
        "U1B"
    FROM game_log
    WHERE "1b_umpire_id" IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        [2b_umpire_id],
        "U2B"
    FROM game_log
    WHERE [2b_umpire_id] IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        [3b_umpire_id],
        "U3B"
    FROM game_log
    WHERE [3b_umpire_id] IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        lf_umpire_id,
        "ULF"
    FROM game_log
    WHERE lf_umpire_id IS NOT NULL

UNION

    SELECT
        game_id,
        NULL,
        rf_umpire_id,
        "URF"
    FROM game_log
    WHERE rf_umpire_id IS NOT NULL

UNION

    SELECT
        game_id,
        v_name,
        v_manager_id,
        "MM"
    FROM game_log
    WHERE v_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        h_name,
        h_manager_id,
        "MM"
    FROM game_log
    WHERE h_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_pitcher_id,
        "AWP"
    FROM game_log
    WHERE winning_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score < v_score THEN h_name
            ELSE v_name
            END,
        losing_pitcher_id,
        "ALP"
    FROM game_log
    WHERE losing_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        saving_pitcher_id,
        "ASP"
    FROM game_log
    WHERE saving_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_rbi_batter_id,
        "AWB"
    FROM game_log
    WHERE winning_rbi_batter_id IS NOT NULL

UNION

    SELECT
        game_id,
        v_name,
        v_starting_pitcher_id,
        "PSP"
    FROM game_log
    WHERE v_starting_pitcher_id IS NOT NULL

UNION

    SELECT
        game_id,
        h_name,
        h_starting_pitcher_id,
        "PSP"
    FROM game_log
    WHERE h_starting_pitcher_id IS NOT NULL;
"""

template = """
INSERT INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "O{num}"
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL

UNION

    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "D" || CAST({hv}_player_{num}_def_pos AS INT)
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL;
"""

run_command(c18)
run_command(c19)

for hv in ["h","v"]:
    for num in range(1,10):
        query_vars = {
            "hv": hv,
            "num": num
        }
        run_command(template.format(**query_vars))

In [16]:
tables = [
    "game_log",
    "park_codes",
    "team_codes",
    "person_codes"
]

for t in tables:
    c = '''
    DROP TABLE {}
    '''.format(t)
    
    run_command(c)

show_tables()

name   type
0               park  table
1             league  table
2             person  table
3               team  table
4               game  table
5    team_appearance  table
6    appearance_type  table
7  person_appearance  table